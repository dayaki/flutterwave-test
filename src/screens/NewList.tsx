import React, { useContext } from 'react';
import { View, Alert } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { useForm, Controller } from 'react-hook-form';
import {
  BackButton,
  Button,
  FormInput,
  PageTitleText,
  ProductProps,
} from '../common';
import { loginStyles as styles } from './styles';
import { addProduct, AppContext, deleteProduct, updateProduct } from '../store';
import { NavigationProps } from '../navigation';

const NewList = ({ navigation, route }: NavigationProps) => {
  const product = route.params?.product;
  const { dispatch } = useContext(AppContext);
  const {
    control,
    handleSubmit,
    formState: { errors },
  } = useForm({
    defaultValues: {
      name: product ? product.name : '',
      stock: product ? product.stock : '',
      price: product ? product.price : '',
      desc: product ? product.desc : '',
    },
  });

  const onDelete = () => {
    Alert.alert('Delete Product?', 'Sure you want to delete this product.', [
      {
        text: 'Cancel',
        style: 'cancel',
      },
      {
        text: 'Delete',
        onPress: () => {
          dispatch(deleteProduct(product));
          navigation.goBack();
        },
      },
    ]);
  };

  const onSubmit = ({ name, price, stock, desc }: ProductProps) => {
    const payload = {
      name,
      price,
      stock,
      desc,
      slug: product
        ? product.slug
        : name.toLocaleLowerCase().split(' ').join('-'),
    };
    dispatch(product ? updateProduct(payload) : addProduct(payload));
    navigation.goBack();
  };

  return (
    <KeyboardAwareScrollView
      style={[styles.container, styles.paddedContainer]}
      contentContainerStyle={styles.wrapper}>
      <BackButton onPress={() => navigation.goBack()} />
      <PageTitleText text="Add New Product" />
      <View style={styles.form}>
        <Controller
          control={control}
          rules={{
            required: true,
          }}
          render={({ field: { onChange, onBlur, value } }) => (
            <FormInput
              placeholder="Product name"
              value={value}
              onBlur={onBlur}
              onChange={onChange}
              hasError={errors.name ? true : false}
              errorMessage="Product name is required"
            />
          )}
          name="name"
        />
        <Controller
          control={control}
          rules={{
            required: 'Total stock is required',
          }}
          render={({ field: { onChange, onBlur, value } }) => (
            <FormInput
              placeholder="Total stock"
              value={value}
              onBlur={onBlur}
              onChange={onChange}
              keyboardType="phone-pad"
              hasError={errors.stock ? true : false}
              errorMessage={errors.stock?.message}
            />
          )}
          name="stock"
        />
        <Controller
          control={control}
          rules={{
            required: 'Price is required',
          }}
          render={({ field: { onChange, onBlur, value } }) => (
            <FormInput
              placeholder="Price"
              value={value}
              onBlur={onBlur}
              onChange={onChange}
              keyboardType="phone-pad"
              hasError={errors.price ? true : false}
              errorMessage={errors.price?.message}
            />
          )}
          name="price"
        />
        <Controller
          control={control}
          rules={{
            required: 'Description is required',
            minLength: {
              value: 3,
              message: 'Product description should be at least 3 digits',
            },
          }}
          render={({ field: { onChange, onBlur, value } }) => (
            <FormInput
              textarea
              placeholder="Description"
              value={value}
              onBlur={onBlur}
              onChange={onChange}
              hasError={errors.desc ? true : false}
              errorMessage={errors.desc?.message}
            />
          )}
          name="desc"
        />
        {product ? (
          <View style={styles.buttons}>
            <Button
              text="Update Product"
              buttonStyle={styles.btn}
              onPress={handleSubmit(onSubmit)}
            />
            <Button
              transparent
              text="Delete Product"
              buttonStyle={styles.btn}
              onPress={onDelete}
              testID="delete-btn"
            />
          </View>
        ) : (
          <Button
            text="Add New Product"
            buttonStyle={styles.button}
            onPress={handleSubmit(onSubmit)}
          />
        )}
      </View>
    </KeyboardAwareScrollView>
  );
};

export default NewList;
