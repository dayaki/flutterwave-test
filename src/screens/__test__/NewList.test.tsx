import React from 'react';
import renderer from 'react-test-renderer';
import NewList from '../NewList';

test('that new product screen renders correctly', () => {
  const newList = renderer
    .create(<NewList route={{ key: 'new_product', name: 'new_product' }} />)
    .toJSON();
  expect(newList).toMatchSnapshot();
});
