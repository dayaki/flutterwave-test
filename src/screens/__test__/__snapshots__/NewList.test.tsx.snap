// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`that new product screen renders correctly 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  contentInset={
    Object {
      "bottom": 0,
    }
  }
  enableAutomaticScroll={true}
  enableOnAndroid={false}
  enableResetScrollToCoords={true}
  extraHeight={75}
  extraScrollHeight={0}
  getScrollResponder={[Function]}
  handleOnScroll={[Function]}
  keyboardDismissMode="interactive"
  keyboardOpeningTime={250}
  keyboardSpace={0}
  onScroll={[Function]}
  resetKeyboardSpace={[Function]}
  scrollEventThrottle={1}
  scrollForExtraHeightOnAndroid={[Function]}
  scrollIntoView={[Function]}
  scrollToEnd={[Function]}
  scrollToFocusedInput={[Function]}
  scrollToPosition={[Function]}
  showsVerticalScrollIndicator={true}
  style={
    Array [
      Object {
        "backgroundColor": "#F7F5F3",
        "flex": 1,
        "paddingHorizontal": 30,
        "paddingTop": 150,
      },
      Object {
        "paddingTop": 50,
      },
    ]
  }
  update={[Function]}
  viewIsInsideTabBar={false}
>
  <View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "flex-start",
          "marginBottom": 30,
          "marginLeft": -10,
          "opacity": 1,
          "paddingHorizontal": 5,
          "paddingVertical": 6,
        }
      }
    >
      <RNSVGSvgView
        bbHeight={32}
        bbWidth={32}
        fill="none"
        focusable={false}
        height={32}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "flex": 0,
              "height": 32,
              "width": 32,
            },
          ]
        }
        width={32}
        xmlns="http://www.w3.org/2000/svg"
      >
        <RNSVGGroup
          fill={null}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M7 15h20a1 1 0 0 1 0 2H7a1 1 0 0 1 0-2Z"
            fill={4278255873}
            propList={
              Array [
                "fill",
              ]
            }
          />
          <RNSVGPath
            d="m7.414 16 8.294 8.292a1.002 1.002 0 0 1-1.416 1.416l-9-9a.999.999 0 0 1 0-1.416l9-9a1.001 1.001 0 1 1 1.416 1.416L7.414 16Z"
            fill={4278255873}
            propList={
              Array [
                "fill",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#010101",
            "fontFamily": "NotoSans-Medium",
            "fontSize": 24,
            "lineHeight": 31,
          },
          undefined,
        ]
      }
    >
      Add New Product
    </Text>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 30,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#F1ECE4",
              "borderRadius": 10,
              "height": 72,
              "justifyContent": "center",
              "marginBottom": 20,
              "width": "100%",
            },
            undefined,
            false,
            undefined,
          ]
        }
      >
        <TextInput
          keyboardType="default"
          numberOfLines={10}
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Product name"
          placeholderTextColor="#474747"
          style={
            Array [
              Object {
                "color": "#010101",
                "fontFamily": "NotoSans-Regular",
                "fontSize": 16,
                "height": 72,
                "lineHeight": 21,
                "paddingLeft": 30,
                "width": "100%",
              },
              undefined,
            ]
          }
          textAlignVertical="top"
          value=""
        />
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#F1ECE4",
              "borderRadius": 10,
              "height": 72,
              "justifyContent": "center",
              "marginBottom": 20,
              "width": "100%",
            },
            undefined,
            false,
            undefined,
          ]
        }
      >
        <TextInput
          keyboardType="phone-pad"
          numberOfLines={10}
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Total stock"
          placeholderTextColor="#474747"
          style={
            Array [
              Object {
                "color": "#010101",
                "fontFamily": "NotoSans-Regular",
                "fontSize": 16,
                "height": 72,
                "lineHeight": 21,
                "paddingLeft": 30,
                "width": "100%",
              },
              undefined,
            ]
          }
          textAlignVertical="top"
          value=""
        />
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#F1ECE4",
              "borderRadius": 10,
              "height": 72,
              "justifyContent": "center",
              "marginBottom": 20,
              "width": "100%",
            },
            undefined,
            false,
            undefined,
          ]
        }
      >
        <TextInput
          keyboardType="phone-pad"
          numberOfLines={10}
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Price"
          placeholderTextColor="#474747"
          style={
            Array [
              Object {
                "color": "#010101",
                "fontFamily": "NotoSans-Regular",
                "fontSize": 16,
                "height": 72,
                "lineHeight": 21,
                "paddingLeft": 30,
                "width": "100%",
              },
              undefined,
            ]
          }
          textAlignVertical="top"
          value=""
        />
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#F1ECE4",
              "borderRadius": 10,
              "height": 72,
              "justifyContent": "center",
              "marginBottom": 20,
              "width": "100%",
            },
            undefined,
            false,
            Object {
              "height": 152,
            },
          ]
        }
      >
        <TextInput
          keyboardType="default"
          multiline={true}
          numberOfLines={10}
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Description"
          placeholderTextColor="#474747"
          style={
            Array [
              Object {
                "color": "#010101",
                "fontFamily": "NotoSans-Regular",
                "fontSize": 16,
                "height": 72,
                "lineHeight": 21,
                "paddingLeft": 30,
                "width": "100%",
              },
              Object {
                "height": 152,
                "width": "94%",
              },
            ]
          }
          textAlignVertical="top"
          value=""
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#DDB582",
            "borderRadius": 10,
            "height": 60,
            "justifyContent": "center",
            "marginTop": 40,
            "opacity": 1,
            "width": "100%",
          }
        }
        testID={null}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#010101",
                "fontFamily": "NotoSans-Medium",
                "fontSize": 14,
                "lineHeight": 21,
                "textAlign": "center",
              },
              undefined,
            ]
          }
        >
          Add New Product
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
